flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/NetworkEnvironmentConfiguration.java

```diff
@@ -105,16 +112,381 @@ public NettyConfig nettyConfig() {
+	}
+
+	/**
+	 * Returns whether the new network buffer memory configuration is present in the configuration
+	 * object, i.e. at least one new parameter is given or the old one is not present.
+	 *
+	 * @param config configuration object
+	 * @return <tt>true</tt> if the new configuration method is used, <tt>false</tt> otherwise
+	 */
+	@SuppressWarnings("deprecation")
+	@VisibleForTesting
+	public static boolean hasNewNetworkConfig(final Configuration config) {
+		return config.contains(TaskManagerOptions.NETWORK_BUFFERS_MEMORY_FRACTION) ||
+			config.contains(TaskManagerOptions.NETWORK_BUFFERS_MEMORY_MIN) ||
+			config.contains(TaskManagerOptions.NETWORK_BUFFERS_MEMORY_MAX) ||
+			!config.contains(TaskManagerOptions.NETWORK_NUM_BUFFERS);
+	}
+
+	/**
+	 * Parses the hosts / ports for communication and data exchange from configuration.
+	 *
+	 * @param configuration configuration object
+	 * @return the data port
+	 */
+	private static int getDataport(Configuration configuration) {
+		final int dataport = configuration.getInteger(TaskManagerOptions.DATA_PORT);
+		ConfigurationParserUtils.checkConfigParameter(dataport >= 0, dataport, TaskManagerOptions.DATA_PORT.key(),
+			"Leave config parameter empty or use 0 to let the system choose a port automatically.");
+
+		return dataport;
+	}
+
+	/**
+	 * Calculates the number of network buffers based on configuration and jvm heap size.
+	 *
+	 * @param configuration configuration object
+	 * @param maxJvmHeapMemory the maximum JVM heap size (in bytes)
+	 * @return the number of network buffers
+	 */
+	private static int calculateNumberOfNetworkBuffers(Configuration configuration, long maxJvmHeapMemory) {
+		final int numberOfNetworkBuffers;
+		if (!hasNewNetworkConfig(configuration)) {
+			// fallback: number of network buffers
+			numberOfNetworkBuffers = configuration.getInteger(TaskManagerOptions.NETWORK_NUM_BUFFERS);
+
+			checkOldNetworkConfig(numberOfNetworkBuffers);
+		} else {
+			if (configuration.contains(TaskManagerOptions.NETWORK_NUM_BUFFERS)) {
+				LOG.info("Ignoring old (but still present) network buffer configuration via {}.",
+					TaskManagerOptions.NETWORK_NUM_BUFFERS.key());
+			}
+
+			final long networkMemorySize = calculateNewNetworkBufferMemory(configuration, maxJvmHeapMemory);
+
+			// tolerate offcuts between intended and allocated memory due to segmentation (will be available to the user-space memory)
+			long numberOfNetworkBuffersLong = networkMemorySize / getPageSize(configuration);
+			if (numberOfNetworkBuffersLong > Integer.MAX_VALUE) {
+				throw new IllegalArgumentException("The given number of memory bytes (" + networkMemorySize
+					+ ") corresponds to more than MAX_INT pages.");
+			}
+			numberOfNetworkBuffers = (int) numberOfNetworkBuffersLong;
+		}
+
+		return numberOfNetworkBuffers;
+	}
+
+	/**
+	 * Parses the configuration to wrapper related components into netty configuration which might be null if
+	 * communication is in the same task manager.
```

Comment:
This is a bit unclear, maybe just:
`Parses Netty configuration from Flink {@link Configuration} which is null if communication is in the same task manager.`

Reply:
I have two concerns of above expression. 
1.  Flink `Configuration` is parsed to generate `NettyConfig`. `Parse Netty Configuration` seems not proper?
2. `NettyConfig` might be null.  But `which is null` indicates Flink `Configuration`?

Or we use `Generates {@link NettyConfig} from parsing Flink {@link Configuration} if communication is in the same task manager`. What do you think?

Reply:
ok, then:
```
/**
  * Generates {@link NettyConfig} from Flink {@link Configuration}.
........
 * @return the netty configuration or {@code null} if communication is in the same task manager
 */
```